core/arch/arm/kernel/secure_partition.c

```diff
@@ -234,14 +231,172 @@ static TEE_Result sp_init_uuid(const TEE_UUID *uuid)
+
+	args->a0 = ctx->sp_regs.x[0];
+	args->a1 = ctx->sp_regs.x[1];
+	args->a2 = ctx->sp_regs.x[2];
+	args->a3 = ctx->sp_regs.x[3];
+	args->a4 = ctx->sp_regs.x[4];
+	args->a5 = ctx->sp_regs.x[5];
+	args->a6 = ctx->sp_regs.x[6];
+	args->a7 = ctx->sp_regs.x[7];
+
 	return res;
 }
 
+static TEE_Result sp_enter_invoke_cmd(struct ts_session *s,
+				      uint32_t cmd __unused)
+{
+	struct sp_ctx *ctx = to_sp_ctx(s->ctx);
+	TEE_Result res = TEE_SUCCESS;
+	uint32_t exceptions = 0;
+	uint64_t cpsr = 0;
+	struct sp_session *sp_s = to_sp_session(s);
+	struct ts_session *sess = NULL;
+	struct thread_ctx_regs *sp_regs = NULL;
+	uint32_t panicked = false;
+	uint32_t panic_code = 0;
+
+	bm_timestamp();
+
+	sp_regs = &ctx->sp_regs;
+	ts_push_current_session(s);
+
+	cpsr = sp_regs->cpsr;
+	sp_regs->cpsr = read_daif() & (SPSR_64_DAIF_MASK << SPSR_64_DAIF_SHIFT);
+
+	exceptions = thread_mask_exceptions(THREAD_EXCP_ALL);
+
+	__thread_enter_user_mode(sp_regs, &panicked, &panic_code);
+	sp_regs->cpsr = cpsr;
+	thread_unmask_exceptions(exceptions);
+
+#ifdef CFG_WITH_VFP
+	thread_user_clear_vfp(&ctx->uctx);
+#endif
+
+	sess = ts_pop_current_session();
+
+	if (panicked) {
+		DMSG("tee_user_ta_enter: SP panicked with code 0x%x",
+		     panic_code);
+		abort_print_current_ts();
+
+		cpu_spin_lock(&sp_s->spinlock);
+		sp_s->state = sp_dead;
+		cpu_spin_unlock(&sp_s->spinlock);
+
+		return TEE_ERROR_TARGET_DEAD;
+	}
+
+	assert(sess == s);
+
+	bm_timestamp();
+	return res;
+}
+
+static void sp_svc_store_registers(struct thread_svc_regs *regs,
+				   struct thread_ctx_regs *sp_regs)
+{
+	sp_regs->x[0] = regs->x0;
+	sp_regs->x[1] = regs->x1;
+	sp_regs->x[2] = regs->x2;
+	sp_regs->x[3] = regs->x3;
+	sp_regs->x[4] = regs->x4;
+	sp_regs->x[5] = regs->x5;
+	sp_regs->x[6] = regs->x6;
+	sp_regs->x[7] = regs->x7;
+	sp_regs->x[8] = regs->x8;
+	sp_regs->x[9] = regs->x9;
+	sp_regs->x[10] = regs->x10;
+	sp_regs->x[11] = regs->x11;
+	sp_regs->x[12] = regs->x12;
+	sp_regs->x[13] = regs->x13;
+	sp_regs->x[14] = regs->x14;
+	sp_regs->x[15] = regs->x15;
+	sp_regs->x[16] = regs->x16;
+	sp_regs->x[17] = regs->x17;
+	sp_regs->x[18] = regs->x18;
+	sp_regs->x[19] = regs->x19;
+	sp_regs->x[20] = regs->x20;
+	sp_regs->x[21] = regs->x21;
+	sp_regs->x[22] = regs->x22;
+	sp_regs->x[23] = regs->x23;
+	sp_regs->x[24] = regs->x24;
+	sp_regs->x[25] = regs->x25;
+	sp_regs->x[26] = regs->x26;
+	sp_regs->x[27] = regs->x27;
+	sp_regs->x[28] = regs->x28;
+	sp_regs->x[29] = regs->x29;
+	sp_regs->x[30] = regs->x30;
+	sp_regs->pc = regs->elr;
+	sp_regs->sp = regs->sp_el0;
+}
+
+static bool sp_handle_svc(struct thread_svc_regs *regs)
+{
+	struct ts_session *ts = NULL;
+	struct sp_ctx *uctx = NULL;
+	struct sp_session *s = NULL;
+
+	ts = ts_get_current_session();
+	uctx = to_sp_ctx(ts->ctx);
+
+	s = uctx->open_session;
+	assert(s);
+
+	sp_svc_store_registers(regs, &uctx->sp_regs);
+
+	regs->x0 = 0;
+	regs->x1 = 0; /* panic */
+	regs->x2 = 0; /* panic code */
+
+	return false;
```

Comment:
Please add a short comment explaining what's going to happen next since it might be a bit hard to follow the flow.

Reply:
Done